File: problems/leetcode/0500/588.cpp
// 588. Design In-Memory File System
// https://leetcode.com/problems/design-in-memory-file-system/description/

using namespace std;

#include <string>
#include <unordered_map>
#include <vector>

vector<string> parsePath(string path) {
  vector<string> tokens;

  int n = path.size();
  string token;
  for (int i = 1; i < n; i++) {
    if (path[i] == '/') {
      tokens.push_back(token);
      token.clear();
    } else {
      token.push_back(path[i]);
    }
  }
  if (!token.empty()) {
    tokens.push_back(token);
  }

  return tokens;
}

class Entry {
public:
  bool directory = false; // true if directory, false if file
  unordered_map<string, Entry *> children;
  string content;

  Entry(bool _directory) { directory = _directory; }

  bool isDirectory() { return directory; }
};

class FileSystem {
public:
  Entry *root = new Entry(true);

  FileSystem() {}

  vector<string> ls(string path) {
    vector<string> tokens = parsePath(path);

    Entry *entry = root;
    for (string &token : tokens) {
      entry = entry->children[token];
    }

    if (!entry->directory) {
      return vector<string>({tokens.back()});
    }

    vector<string> result;
    for (pair<string, Entry *> p : entry->children) {
      result.push_back(p.first);
    }
    sort(result.begin(), result.end());

    return result;
  }

  void mkdir(string path) {
    vector<string> tokens = parsePath(path);

    Entry *entry = root;
    for (string &token : tokens) {
      if (!entry->children[token]) {
        entry->children[token] = new Entry(true);
      }
      entry = entry->children[token];
    }
  }

  void addContentToFile(string filePath, string content) {
    vector<string> tokens = parsePath(filePath);

    Entry *entry = root;
    for (int i = 0; i < tokens.size() - 1; i++) {
      if (!entry->children[tokens[i]]) {
        entry->children[tokens[i]] = new Entry(true);
      }
      entry = entry->children[tokens[i]];
    }
    if (!entry->children[tokens.back()]) {
      entry->children[tokens.back()] = new Entry(false);
    }
    entry = entry->children[tokens.back()];
    entry->content += content;
  }

  string readContentFromFile(string filePath) {
    vector<string> tokens = parsePath(filePath);

    Entry *entry = root;
    for (string &token : tokens) {
      entry = entry->children[token];
    }

    return entry->content;
  }
};

/**
 * Your FileSystem object will be instantiated and called as such:
 * FileSystem* obj = new FileSystem();
 * vector<string> param_1 = obj->ls(path);
 * obj->mkdir(path);
 * obj->addContentToFile(filePath,content);
 * string param_4 = obj->readContentFromFile(filePath);
 */

// 1. Directory & File 관리
// 2. File은 string 내용 가지고 있음

File: problems/leetcode/0500/588/588.py
# 588. Design In-Memory File System
# https://leetcode.com/problems/design-in-memory-file-system/

class File:

    def __init__(self, name: str, content: str = None):
        self.name = name
        self.content = content

    def add(self, content: str):
        self.content += content

class Directory:

    def __init__(self):
        self.children = dict()

    def add(self, key: str, value: Union[File, Directory]):
        if key not in self.children:
            self.children[key] = value

    def exist(self, key: str):
        return key in self.children

    def getNames(self) -> List[str]:
        return [n for n in self.children.keys()]

class FileSystem:

    def __init__(self):
        self.root = Directory()

    def ls(self, path: str) -> List[str]:
        node = self.root

        for n in path.split("/")[1:]:
            if n != "":
                node = node.children[n]

        if isinstance(node, File):
            return [node.name]

        return [n for n in sorted(node.getNames())]

    def mkdir(self, path: str) -> None:
        node = self.root

        for n in path.split("/")[1:]:
            if n not in node.children:
                node.add(n, Directory())
            node = node.children[n]

    def addContentToFile(self, filePath: str, content: str) -> None:
        paths = filePath.split("/")[1:]
        fileName = paths[-1]
        node = self.root

        for n in paths[:-1]:
            node = node.children[n]

        if node.exist(fileName):
            node.children[fileName].add(content)
            return

        node.add(fileName, File(fileName, content))

    def readContentFromFile(self, filePath: str) -> str:
        node = self.root

        for n in filePath.split("/")[1:]:
            node = node.children[n]

        return node.content


# Your FileSystem object will be instantiated and called as such:
# obj = FileSystem()
# param_1 = obj.ls(path)
# obj.mkdir(path)
# obj.addContentToFile(filePath,content)
# param_4 = obj.readContentFromFile(filePath)
